grafana/grafana
# Prometheus: Support utf-8 labels and metric names

**What is this feature?**

It adds UTF-8 label names support in Prometheus data source.

Reference: https://github.com/prometheus/prometheus/issues/13095
Proposal: https://github.com/prometheus/proposals/blob/main/proposals/2023-08-21-utf8.md
Presentation: https://docs.google.com/presentation/d/1hSfMOwUF74GoD7LG5MtOaLChleLKj9jFuRibTtwJM9k/edit#slide=id.g2f665c27ee0_0_195

**How can you test it?**
- I provided the devenv `prometheus-utf8` for this purpose.
- Just run `make devenv sources=prometheus`
- It will fetch the latest prometheus and build up our metric sources that will be scraped by our prom
- In general you should be able to get the latest prometheus. But if not just stop all prometheus containers. And run `docker image prune -a` https://docs.docker.com/reference/cli/docker/image/prune/ And run devenv command again.
- We have a normal metric with utf8 labels `a_utf8_http_requests_total`
- And we have a utf8 metric `a.utf8.metric 🤘` with all sorts of labels

**Why do we need this feature?**

To support utf-8 label names in prometheus data source

**Who is this feature for?**

Prometheus users who has utf-8 label names

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/91253
Fixes https://github.com/grafana/grafana/issues/74875


**Missing Parts**
- [x] In the table view, clicking "Filter for/out value" button doesn't update the promql properly.
- [x] In the raw view "Copy to clipboard" doesn't give utf-8 labels in quotes
- [x] Label name query returns the labels but while appending them to promql we don't add quotes around them.
- [x] When selecting a utf8 label and switching to builder mode, utf8 label doesn't show up
- [x] In explanation utf8 labels are not wrapped in quotes
- [x] When switching to builder mode with an utf8 metric name builder doesn't show the metric name i.e. `{"utf8.metric 🤘", "label with 😱"="fear not"}`
- [x] With an utf8 metric, raw mode shows the query in legacy way `utf8.metric{a_legacy_label="legacy"}` it should be `{"utf8.metric", a_legacy_label="legacy"}`
- [x] a utf8 label's value should be queried with `label_values`. We should check the whole variable editor.
- [x] utf8 metrics and labels should be supported in explore metrics
- [x] Cannot query the utf8 label values https://github.com/prometheus/prometheus/pull/15399
- [x] Cannot get labels when grouped by utf8 label https://github.com/prometheus/prometheus/issues/15470
- [x] `sum by ("label with space")(rate(fakedata_highcard_http_requests_total{"label with space"="default"}[1m]))` for this promql table view or raw view couldn't show the label name. We don't receive the label name from backend https://github.com/prometheus/prometheus/pull/15531
- [x] When querying utf8 metric labels it requests all labels. It should request only labels for the metric.
- [ ] It should work with scopes too. `promQLScope`, `scopeFilters`, `groupByVariable`, `enableScopesInMetricsExplore` are the feature toggle needs to be enabled. Waiting for prometheus `v0.300.2` to be released

**Special notes for your reviewer:**

Please check that:
